rust/src/sdk/api.rs

```diff
@@ -87,85 +84,174 @@ macro_rules! PrimModel {
+/// Ignores null pointer `client`s and triggers the callback immediately. Triggers the callback
+/// with one of the following error codes in case of a [`Client`] panic or error:
+/// - `-1`: client didn't start due to null pointer
+/// - `0`: no error
+/// - `1`: client panicked due to unexpected/unhandled error
+/// - `2`: client stopped due to error [`ParticipantInitErr`]
+/// - `3`: client stopped due to error [`ParticipantErr`]
+/// - `4`: client stopped due to error [`DeserialiseErr`]
+/// - `5`: client stopped due to error [`GeneralErr`]
+///
+/// # Safety
+/// The method dereferences from the raw pointer arguments. Therefore, the behavior of
+/// the method is undefined if the arguments don't point to valid objects.
+///
+/// If the callback is triggered because of a panicking client, it is the users responsibility
+/// to not access possibly invalid state and to drop the client. If certain parts of the state
+/// can be guaranteed to be valid, they may be read before dropping.
+///
+/// [errors]: #errors
+/// [`ParticipantInitErr`]: ../../client/enum.ClientError.html#variant.ParticipantInitErr
+/// [`ParticipantErr`]: ../../client/enum.ClientError.html#variant.ParticipantErr
+/// [`DeserialiseErr`]: ../../client/enum.ClientError.html#variant.DeserialiseErr
+/// [`GeneralErr`]: ../../client/enum.ClientError.html#variant.GeneralErr
+pub unsafe extern "C" fn run_client(
+    client: *mut FFIClient,
+    callback: unsafe extern "C" fn(*mut c_void, c_int),
+    state: *mut c_void,
+) {
+    if client.is_null() {
+        return unsafe {
+            // safe if the `callback` is sound
+            callback(state, -1_i32 as c_int)
+        };
     }
-    let input = &Input {
-        _round_started: client.has_new_coord_pk_since_last_check,
-        _participant_initialized: true,
-        _model_cached: client.cached_model.is_some(),
-    } as *const _ as *const c_void;
+    let (runtime, client) = unsafe {
+        // safe if the raw pointer `client` comes from a valid allocation of a `FFIClient`
+        (&(*client).runtime, &mut (*client).client)
+    };
+    let code = match panic::catch_unwind(
+        // even though `&mut Client` is `!UnwindSafe` we can assert this because the user will be
+        // notified about a panic immediately to be able to safely act accordingly
+        panic::AssertUnwindSafe(|| runtime.handle().block_on(client.start())),
+    ) {
+        Ok(Ok(_)) => 0_i32,
+        Err(_) => 1_i32,
+        Ok(Err(ClientError::ParticipantInitErr(_))) => 2_i32,
+        Ok(Err(ClientError::ParticipantErr(_))) => 3_i32,
+        Ok(Err(ClientError::DeserialiseErr(_))) => 4_i32,
+        Ok(Err(ClientError::GeneralErr)) => 5_i32,
+    } as c_int;
     unsafe {
-        // safe if the `callback` is sound and the same definition and layout is used for `Input`
-        // across the FFI-boundary by the caller
-        callback(state, input)
+        // safe if the `callback` is sound
+        callback(state, code)
```

Comment:
The doc states:

> The callback will be triggered when the client must be stopped because of a panic or error

But it is actually called even when there's no error.

Reply:
this callback is currently unreachable because `client.start()` will only ever stop if an error happens. we could either remove this specific call of the callback or update the docs?

Reply:
i updated the docs for now.